src/main/java/com/minecolonies/colony/buildings/BuildingBuilder.java

```diff
@@ -96,23 +120,131 @@ public String getJobName()
      */
     @NotNull
     @Override
-    public AbstractJob createJob(CitizenData citizen)
+    public AbstractJob createJob(final CitizenData citizen)
     {
         return new JobBuilder(citizen);
     }
 
     /**
+     * Get the needed resources for the current build.
+     * @return a new Hashmap.
+     */
+    public Map<Block, Integer> getNeededResources()
+    {
+        return new HashMap<>(neededResources);
+    }
+
+    /**
+     * Add a new resource to the needed list.
+     * @param res the resource.
+     * @param amount the amount.
+     */
+    public void addNeededResource(Block res, int amount)
+    {
+        int preAmount = 0;
+        if(this.neededResources.containsKey(res))
+        {
+            preAmount = this.neededResources.get(res);
+        }
+        this.neededResources.put(res, preAmount + amount);
+        this.markDirty();
+    }
+
+    /**
+     * Reduce a resource of the needed list.
+     * @param res the resource.
+     * @param amount the amount.
+     */
+    public void reduceNeededResource(Block res, int amount)
+    {
+        int preAmount = 0;
+        if(this.neededResources.containsKey(res))
+        {
+            preAmount = this.neededResources.get(res);
+        }
+
+        if(preAmount - amount <= 0)
+        {
+            this.neededResources.remove(res);
+        }
+        else
+        {
+            this.neededResources.put(res, preAmount - amount);
+        }
+        this.markDirty();
+    }
+
+    /**
+     * Resets the needed resources completely.
+     */
+    public void resetNeededResources()
+    {
+        neededResources = new HashMap<>();
+        this.markDirty();
+    }
+
+    @Override
+    public void readFromNBT(@NotNull final NBTTagCompound compound)
+    {
+        super.readFromNBT(compound);
+        final NBTTagList neededResTagList = compound.getTagList(TAG_RESOURCE_LIST, Constants.NBT.TAG_COMPOUND);
+        for (int i = 0; i < neededResTagList.tagCount(); ++i)
+        {
+            NBTTagCompound neededRes = neededResTagList.getCompoundTagAt(i);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'neededRes' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)